ui: React router version upgrade

koorosh opened the pull request:
Migration to the latest version of 'react-router' version (v4) requires to fix following
breaking changes:
- change imports for Link, Route components from 'react-router' to 'react-router-dom' library;
- 'push' and 'replace' history actions are called from 'History' component
instead of 'Route'
- Accessing to search params is changed, now it is accessed from 'history.location.search'.
- Components which passed to <Route> component is wrapped by 'withRoute' HOC. It is
required to properly re-render component when routing properties change;
- Access to Route Params are changed as well, it is now accessed from 'match' property
which is passed from <Route>
- Nested routes are not supported anymore. Refactored to use `flat`
representation of routes;
- Relative paths aren't supported, changed to absolute paths;
- history is provided as an argument to `createAdminUIStore` function.
It decouples responsibility of history instance creation from story initialization.
Also it makes easier to test store and supply test instance of history.
- Refactored `AnalyticsSync#pushPage` method to avoid typing errors and properly
iterate over search params.
- `AdminUIState.routing` is renamed to 'AdminUIState.router' as described in
'connected-react-router' documentation. We don't use this field in code so it
is safe and doesn't affect any functionality.

koorosh committed:
ui: Upgrade React version and react-dependent libraries

With current changes, react is upgraded to latest version (16.12.0) and
following libraries are upgraded as well: react-redux, redux, redux-saga.
These upgrades of major versions has required to do some code adjustments
to properly migrate to newest versions.
- redux. Provide valid type arguments for Dispatch type.
- redux-saga. Changed import and usage of 'delay' function.
- react-redux. `connect` function failed to correctly resolve returned type
for `mapDispatchToProps` argument. As result it is provided as a function
instead of object argument. This hack allows implicitly resolve types.
- `CachedDataReducer` class is extended with one more optional type argument
to specify the list of allowed `actionNamespace` literals instead of `string`.
It was necessary to have strictly defined CachedDataReducer interface.
- `normalizeConnectedComponent` function is used to do simple mapping from
react's ExoticComponent to valid react component. This resolves issue of
providing ConnectedComponents (which are ExoticComponent's, not regular react
components) to Route component which has strict type validation of component types.

Release note: None

koorosh committed:
ui: Upgrade React version and react-dependent libraries

With current changes, react is upgraded to latest version (16.12.0) and
following libraries are upgraded as well: react-redux, redux, redux-saga.
These upgrades of major versions has required to do some code adjustments
to properly migrate to newest versions.
- redux. Provide valid type arguments for Dispatch type.
- redux-saga. Changed import and usage of 'delay' function.
- react-redux. `connect` function failed to correctly resolve returned type
for `mapDispatchToProps` argument. As result it is provided as a function
instead of object argument. This hack allows implicitly resolve types.
- `CachedDataReducer` class is extended with one more optional type argument
to specify the list of allowed `actionNamespace` literals instead of `string`.
It was necessary to have strictly defined CachedDataReducer interface.
- `normalizeConnectedComponent` function is used to do simple mapping from
react's ExoticComponent to valid react component. This resolves issue of
providing ConnectedComponents (which are ExoticComponent's, not regular react
components) to Route component which has strict type validation of component types.

Release note: None

koorosh committed:
ui: Fix version of enzyme-react-adapter library

Release note: None

koorosh committed:
ui: Fix version of enzyme-react-adapter library

Release note: None

koorosh committed:
ui: Fix "Maximum call stack size" issue with react-helmet library

- Issue: https://github.com/nfl/react-helmet/issues/373
- Workaround: https://github.com/nfl/react-helmet/issues/373#issuecomment-392650715
- The issue occurs when tests are executed with mounting root App.

Release note: None

koorosh committed:
ui: Add tests for /statement(s) routes

Release note: None

koorosh committed:
Upgrade react-router lib and its dependencies

koorosh committed:
ui: Refactor imports from 'react-router' to 'react-router-dom' library

Migration to the latest version of 'react-router' version (v4) requires to fix
breaking changes:
- change imports for Link, Route components.
- 'push' and 'replace' history actions are called from 'Histry' component
instead of 'Route'
- Accessing to search params are changed, accessed from 'history.location.search'.
- Components which passed to <Route> component is wrapped by 'withRoute'
- Access to Route Params are changed as well, it is now accessed from 'match' property
which is passed from <Route>

Release note: None

koorosh committed:
ui: Fix app routing

With latest version of 'react-router' there are following changes
required:
- Nested routes are not supported anymore. Refactored to use `flat`
representation of routes;
- Relative paths aren't supported, changed to absolute paths;

Release note: None

koorosh committed:
ui: Fix routing integration with redux store

- history is provided as an argument to `createAdminUIStore` function.
It decouples responsibility of history instance creation from story initialization.
Also it makes easier to test store and supply test instance of history.
- Refactored `AnalyticsSync#pushPage` method to avoid typing errors and properly
iterate over search params.
- `AdminUIState.routing` is renamed to 'AdminUIState.router' as described in
'connected-react-router' documentation. We don't use this field in code so it
is safe and doesn't affect any functionality.

Release note: None

koorosh committed:
ui: Fix routing handling for Databases view

- Refactor navigation handling from 'DatabaseListNav' component.
- Use absolute paths to navigate instead of relative.

Release note: None

koorosh committed:
ui: Fix usage of correct types and react warnings

- Add `key` prop to react components.
- Use 'ILiveness' interface across all app to avoid
typing errors.
- Remove unnecessary definitions: createLocation.d.ts,
createPath.d.ts. They are already defined in 'react-router-dom'
library.

Release note: None

koorosh committed:
ui: Import analytics on application bootstrap

Analytics is moved back on the application initialization.
It makes sure:
- it is initialized only once.
- <App /> component doesn't have additional dependency on
analytics.

Release note: None

koorosh committed:
ui: Fix test data generation for statements tests

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/44214)
<!-- Reviewable:end -->

koorosh committed:
ui: Resolve conflicts after rebasing against master

- tests command is disabled for now because current state doesn't
contain all fixes to run tests without failure.
- Decommissioned Node History page was added so code has to be adjusted.

Release note: None

koorosh committed:
ui: Resolve conflicts after rebasing against master branch

- Wrap NodeCanvasContainer and statementsPage components with `bindActionCreators`
- Add route for DecommissionedNodeHistory page

Release note: None

koorosh committed:
ui: Fix conflicts after rebasing against master branch

Release note: None

koorosh committed:
ui: Get rid of useless function wrappers

It is not necessary to wrap action creator object
with function when connecting it to component.

Release note: None

koorosh committed:
ui: Fix roting paths for Network and DecommissionedNodeList pages

Release note: None

koorosh committed:
ui: Fix Active links highlighting in navigation side panel

It is required to have more granular way to specify if current path
is active or not. To do so, `ignoreFor` option added which contains
list of paths to exclude.

Release note: None

koorosh committed:
ui: Fix parsing router params for Network and NodeFilter components

Release note: None

koorosh committed:
ui: Apply default style classes for H2 heading

Release note: None

dhartunian commented on pkg/ui/src/app.tsx:
```diff
@@ -78,140 +78,114 @@ export const App: React.FC<AppProps> = (props: AppProps) => {
-                <IndexRedirect to={`/node/:${nodeIDAttr}`}/>
-                <Redirect from="logs" to={`/node/:${nodeIDAttr}/logs`}/>
-              </Route>
+      <ConnectedRouter history={history}>
```
Some of the tests in `app.spec.tsx` fail for me, can you take a look?

I noticed that the links in the Statements page in particular were broken too. The individual statements wouldn't link to their details pages and the dropdown wouldn't do anything.

Koorosh replied:
Fixed. The problem was in wrong ordering of routes, more general route path has been placed before more specific one.

dhartunian reviewed (changes requested):
Great work @Koorosh! Just have a few small issues to resolve before we merge. Thanks!

Koorosh commented on pkg/ui/src/views/cluster/containers/clusterOverview/index.tsx:
```diff
@@ -188,7 +188,7 @@ class ClusterSummary extends React.Component<ClusterSummaryProps, {}> {
     }
 
-    return <section className="cluster-summary" children={children} />;
+    return <section className="cluster-summary" children={React.Children.toArray(children)} />;
```
React component requires `key` property for every child element in an array. 
`React.Children.toArray` applies `key` property for every element in the collection, so it is not necessary to manually iterate over collection.

dhartunian replied:
Neat! Didn't know that. Thanks!

Koorosh commented on pkg/ui/src/views/app/components/layoutSidebar/index.tsx:
```diff
@@ -10,29 +10,49 @@
+    const { pathname } = this.props.location;
+    const { activeFor, ignoreFor = []} = this.routes.find(route => route.path === path);
+    return [...activeFor, path].some(p => {
+      const isMatchedToIgnoredPaths = ignoreFor.some(ignorePath => pathname.startsWith(ignorePath));
```
`ignoreFor` is necessary to manage one tricky condition which we have with `Network Latency` view.
`Network Latency` page has `/reports/network` path (since this view was under `Debug` page for a while).

`Debug` path --> `path: "/debug"` and activeFor: ["**/reports**", ...]
`Network` path --> `path: "/reports/network"` and activeFor: ["**/reports**/network"]

`/reports` path (for Debug page) includes "/reports/network" as well. As result two links (Debug and Network Latency) are set as active. To avoid this, I decided to extend this configuration with `ignoreFor` option.

I totally agree with you ... this PR isn't a good place to introduce these changes.

dhartunian replied:
Gotcha. I'll make a separate ticket to clean up the path discrepancy later. Thanks for the detailed explanation!

Koorosh commented on pkg/ui/package.json:
```diff
@@ -10,9 +10,11 @@
     "create-react-context": "^0.3.0",
     "highlight.js": "^9.10.0",
     "lodash": "^4.17.4",
+    "mini-create-react-context": "^0.3.2",
```
It is workaround for issue with bundling lib with in `webpack.vendor.js` and the problem that it cannot resolve ES modules in dependencies (`react-router-dom` is imported as ESM and its dependencies as well (`react-router -> mini-create-react-context`

In fact we have the same problem as described here: https://github.com/ReactTraining/react-router/issues/6739#issuecomment-499659796

Including  `mini-create-react-context` library to `dependencies` list allows webpack to directly bundle this dependency and avoid failed resolution for this module.

I tried to do the same trick in `webpack.vendor.js` line 46 but with no luck :(
```
// analytics-node is an es6 module and we must run it through babel. As
// of 8/25/2017 there appears to be no better way to run babel only on
// the specific packages in node_modules that actually use ES6.
// https://github.com/babel/babel-loader/issues/171
exclude: /node_modules\/(?!analytics-node)/,
```

dhartunian replied:
Another great argument to get rid of the separate `webpack.vendor.js`. Thanks for making this work. I'll make a separate issue to revisit the vendor js generation process.

koorosh committed:
ui: Provide `exact` match for Routes

`exact` property is added to Routes to avoid
dependencies on order of Routes declarations.

Release note: None

dhartunian reviewed (approved):
LGTM

dhartunian commented:
bors r+
